File: face-recognition/microsoft-cognitive-services-face/MicrosoftCognitiveServicesFaceFaceRecognition.java
// input is a string holding the path to the image file.

public class Rectangle {
    public int top, left, width, height;
}

public class Accessory {
    public String type;
    public float confidence;
}

public class Blur {
    public String blurLevel;
    public float value;
}

public class Emotion {
    public short anger, contempt, disgust, fear, happiness, neutral, sadness, surprise;
}

public class Exposure {
    public String exposureLevel;
    public float value;
}

public class FacialHair {
    public short moustache, beard, sideburns;
}

public class HairColor {
    public String color;
    public float confidence;
}

public class Hair {
    public float bald;
    public boolean invisible;
    public List<HairColor> hairColor;
}

public class HeadPose {
    public float pitch, roll, yaw;
}

public class Makeup {
    public boolean eyeMakekup, lipMakeup;
}

public class Noise {
    public String noiseLevel;
    public Float value;
}

public class Occlusion {
    public boolean foreheadOccluded, eyeOccluded, mouthOccluded;
}

public class FaceAttributes {
    public List<Accessory> accessories;
    public short age;
    public Blur blur;
    public Emotion emotion;
    public Exposure exposure;    
    public FacialHair facialHair;
    public String gender;
    public String glasses;
    public Hair hair;
    public HeadPose headPose;
    public Makeup makeup;
    public Noise noise;
    public Occlusion occlusion;
    public short smile;
}

public class FaceRecognition {
    public UUID faceId;
    public Rectangle faceRectangle;
    public FaceAttributes faceAttributes;
}

public class FaceRecognitions extends ArrayList<FaceRecognition> {}

Path path = Paths.get(input);
byte[] bytes = Files.readAllBytes(path);

String subscriptionKey = Credentials.INSTANCE.getMicrosoftFaceKey();

String host = "https://westcentralus.api.cognitive.microsoft.com";
String endpoint = "/face/v1.0/detect";
String base = host.concat(endpoint);

String faceAttributes =
    "age,gender,headPose,smile,facialHair,glasses,emotion,hair,makeup,occlusion,accessories,blur,exposure,noise";

String url = new URIBuilder(base)
    .setParameter("returnFaceId", "true")
    .setParameter("returnFaceLandmarks", "false")
    .setParameter("returnFaceAttributes", faceAttributes)
    .toString();

String json = Unirest.post(url)
    .header("Content-Type", "application/octet-stream")
    .header("Ocp-Apim-Subscription-Key", subscriptionKey)
    .body(bytes)
    .asString()
    .getBody();

Gson gson = new Gson();

FaceRecognitions faceRecognitions = gson.fromJson(json, FaceRecognitions.class);

File: face-detection/microsoft-cognitive-services-face/MicrosoftCognitiveServicesFaceFaceRecognition.kt
// input is a File object pointing to the input image file.

data class Accessory(val type: String, val confidence: Float)

data class Blur (val blurLevel: String, val value: Float)

data class Emotion(
    val anger : Short,
    val contempt : Short,
    val disgust : Short,
    val fear : Short,
    val happiness : Short,
    val neutral : Short,
    val sadness : Short,
    val surprise : Short)

data class Exposure(
    val exposureLevel: String,
    val value: Float)

data class FacialHair(
    val moustache: Short,
    val beard: Short,
    val sideburns: Short)

data class Hair(
    val bald: Float,
    val invisible: Boolean,
    val hairColor: List<HairColor>)

data class HairColor(
    val color: String,
    val confidence: Float)

data class HeadPose(
    val pitch: Float,
    val roll: Float,
    val yaw: Float)

data class Makeup(
    val eyeMakekup: Boolean,
    val lipMakeup: Boolean)

data class Noise(
    val noiseLevel: String,
    val value: Float)

data class Occlusion(
    val foreheadOccluded: Boolean,
    val eyeOccluded: Boolean,
    val mouthOccluded: Boolean)

data class Rectangle(
    val top: Int,
    val left: Int,
    val width: Int,
    val height: Int)

data class FaceAttributes(
    val accessories: List<Accessory>?,
    val age: Short?,
    val blur: Blur?,
    val emotion: Emotion?,
    val exposure: Exposure?,
    val facialHair: FacialHair?,
    val gender: String?,
    val glasses: String?,
    val hair: Hair?,
    val headPose: HeadPose?,
    val makeup: Makeup?,
    val noise: Noise?,
    val occlusion: Occlusion?,
    val smile: Short?)

data class FaceRecognition(
    val faceId: UUID,
    val faceRectangle: Rectangle,
    val faceAttributes: FaceAttributes)

class FaceRecognitions : ArrayList<FaceRecognition>()

val host = "https://westcentralus.api.cognitive.microsoft.com"
val endpoint = "/face/v1.0/detect"
val base = host + endpoint

val faceAttributes = "age,gender,headPose,smile,facialHair,glasses,emotion,hair,makeup,occlusion,accessories,blur,exposure,noise"

val url = URIBuilder(base)
    .setParameter("returnFaceId", "true")
    .setParameter("returnFaceLandmarks", "false")
    .setParameter("returnFaceAttributes", faceAttributes)
    .toString()

val bytes = input.readBytes()

val json = Unirest.post(url)
    .header("Content-Type", "application/octet-stream")
    .header("Ocp-Apim-Subscription-Key", subscriptionKey)
    .body(bytes)
    .asString()
    .getBody()

val gson = Gson()
val faceRecognitions = gson.fromJson(json, FaceRecognitions::class.java)
